Rework room freeze and implement unfreezing the room

babolivier opened the pull request:
A feature we currently have in the `RoomAccessRules` module is the ability to "freeze" room when the last admin has left it (https://github.com/matrix-org/synapse-dinsic/pull/59). Freezing the room means changing the room's rules in a way that prevents anyone from sending messages in the room except for leaving or taking over and unfreezing the room.

The current solution for freezing a room doesn't allow the room to be unfrozen. This is because the Matrix spec explicitly forbids a user to change their own power level, or someone else's, to a higher value than their current one. As per https://spec.matrix.org/unstable/rooms/v1/#authorization-rules:

```
10. If type is m.room.power_levels:
  [...]
  4. For each entry being added, changed or removed in both the events and users keys:
    1. If the current value is higher than the sender’s current power level, reject.
    2. If the new value is higher than the sender’s current power level, reject.
```

This means that from a Matrix perspective the room is effectively broken, because even if the power level event allows anyone to send a new `m.room.power_levels` event, whoever sends it won't be able to change its content. This means unfreezing can't work.

This means the room freezing feature needs to be changed in the following way:

* a new custom state event, `io.element.room.frozen`, is defined. Its content is `{"frozen": [...]}`, with `frozen` a boolean indicating whether the room is frozen, defaulting to `false`.
* freezing a room updates the power levels by only changing the `user_default` key to 100 (rather than sending a brand new event with every permission up'd to 100), and removing any user whose PL isn't specifically set to 100 in the `users` object (in order to not prevent them from being able to unfreeze and take over the room
* freezing a room also changes the value for `io.element.room.frozen`'s `frozen` key in the room's state, setting it to `true`
* the `RoomAccessRules` module forbids new events (except for leaving or unfreezing the room - see below for how the latter is defined) if `frozen` is `true`
* new client UI changes are introduced, triggering if `frozen` is `true` in the room's state:
  * showing the user they can't send new messages in the room
  * users aren't showed as `Admin` in the room list - setting `user_default` to 100 will have the side-effect of showing every user as admin in the room list, and I think clients should hide that to avoid any confusion around the current state of the room.

Additionally, freezing the room would set its join rules to `invite`.

This would then allow unfreezing a room by sending a new power level events that:

* gives the user unfreezing the room power level 100
* reset the `user_default` key back to 0

Here's how this PR implements it:

* when the last admin leaves the room, the `RoomAccessRules` module sends a `io.element.room.frozen` event; this in turns triggers a power levels update to be sent
* if a user sends a `io.element.room.frozen` event manually, the module updates the power levels event accordingly
* when a room is frozen any event is forbidden, except for events that allow someone to leave the room, or for events that relate to the freezing/unfreezing of the room

Future work (out of scope for this PR) include:

* promoting the user(s) with the highest PL to admin once the last admin leaves the room - and only freezing once all other users in the room have the default PL
* splitting this work out in a new module (which would require migrating the `ThirdPartyEventRules` interface to the new generic modules system first, since we'd then need to run two modules with this interface for synapse-dinsic)

babolivier committed:
Implement the new solution for freezing (and unfreezing) a room

babolivier committed:
Rebuild events with a brand new context

babolivier committed:
Refactor updating power levels, and tests

babolivier committed:
Don't update PLs on noops

babolivier committed:
Lint

babolivier committed:
Lint

babolivier committed:
Remove unused imports

babolivier commented:
Failing CI is blocked on https://github.com/matrix-org/synapse-dinsic/pull/99

richvdh commented:
> freezing a room updates the power levels by only changing the user_default key to 100

so every user gets a PL of 100? this seems a curious choice.

richvdh commented on synapse/handlers/message.py:
```diff
@@ -1416,9 +1416,11 @@ async def _rebuild_event_after_third_party_rules(
         # auth events.
+        prev_event_ids = await self.store.get_prev_events_for_room(builder.room_id)
+
         event = await builder.build(
-            prev_event_ids=original_event.prev_event_ids(),
-            auth_event_ids=original_event.auth_event_ids(),
+            prev_event_ids=prev_event_ids,
+            auth_event_ids=None,
```
why is this changing? it seems to now do the exact opposite to what the comment says?

babolivier replied:
Ugh, looks like I forgot to finish that bit up. I also wanted to PR it to mainline so it doesn't end up producing another divergence, I've just done that here: https://github.com/matrix-org/synapse/pull/10316

richvdh replied:
the comment still needs updating for this branch.

babolivier commented:
> so every user gets a PL of 100? this seems a curious choice.

It seems like the only way to let someone name themselves as the new admin of the room (in the process of unfreezing it)

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -394,6 +405,130 @@ async def check_visibility_can_be_modified(
+        if not frozen:
+            # We're unfreezing the room: enforce the right value for the power levels so
+            # the room isn't in a weird/broken state afterwards.
+            users = power_levels_content.get("users", {})
+            users[event.sender] = 100
+            power_levels_content["users"] = users
```
```suggestion
            users = power_levels_content.setdefault("users", {})
            users[event.sender] = 100
```

babolivier replied:
Oooh I didn't know (or more likely forgot) `users` was a reference here, thanks!

richvdh replied:
it's not the most critical thing, but it would be nice to clean this up.

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -394,6 +405,130 @@ async def check_visibility_can_be_modified(
+            # except users_default is 100 to allow any user to unfreeze the room.
+            power_levels_content["users_default"] = 100
+
+            # Just to be safe, also delete all users that don't have a power level of
+            # 100, in order to prevent anyone from being unable to unfreeze the room.
+            users = {}
+            for user, level in power_levels_content["users"].items():
+                if level == 100:
+                    users[user] = level
+            power_levels_content["users"] = users
```
this means that anyone who previously had a `moderator` PL will lose it after a freeze/unfreeze?

babolivier replied:
Yes, but this isn't an issue since future work will make it so if the last admin leaves but there are mods in the room, they get promoted to admin; so at that point there will not be any moderator left.

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -459,7 +594,10 @@ async def _on_membership_or_invite(
-            True if the event can be allowed, False otherwise.
+            A boolean indicating whether the event is allowed, or a dict if the event is
+            allowed but the state of the room has been modified (i.e. the room has been
+            frozen). This is because returning a dict of the event forces Synapse to
+            rebuild it, which is needed if the state of the room has changed.
```
... not according to the docstring on `check_event_allowed` ?

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -484,6 +622,9 @@ async def _on_membership_or_invite(
             ):
                 await self._freeze_room_if_last_admin_is_leaving(event, state_events)
+                if ret:
+                    # Return an event dict to force Synapse into rebuilding th event.
```
th

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -535,92 +676,19 @@ async def _freeze_room_if_last_admin_is_leaving(
             }
         )
 
+        return
```
this is redundant?

richvdh commented:
> > so every user gets a PL of 100? this seems a curious choice.
> 
> It seems like the only way to let someone name themselves as the new admin of the room (in the process of unfreezing it)

yeah fair enough. It just feels a bit surprising that *anyone* can now come in and take over the room. But maybe that is the intention?

babolivier commented:
> > > so every user gets a PL of 100? this seems a curious choice.
> > 
> > 
> > It seems like the only way to let someone name themselves as the new admin of the room (in the process of unfreezing it)
> 
> yeah fair enough. It just feels a bit surprising that _anyone_ can now come in and take over the room. But maybe that is the intention?

Yep exactly (see https://github.com/matrix-org/matrix-dinsic/issues/703 which states "any existing member can take the administration of the room")

babolivier committed:
Fix doc and types for check_event_allowed

babolivier committed:
Merge branch 'dinsic' into babolivier/freeze_unfreeze

babolivier committed:
Copy the content of the PL events before modifying it

Otherwise we end up poisoning the event cache since we're getting a
reference when assigning (and so we're modifying the same content that's
in the cache for that event).

richvdh commented on synapse/third_party_rules/access_rules.py:
```diff
@@ -334,7 +334,8 @@ async def check_event_allowed(
 
         Returns:
-            True if the event can be allowed, False otherwise.
+            True if the event can be allowed, False otherwise, or a dictionary if the
```
```suggestion
            True if the event should be allowed, False if it should be rejected, or a dictionary if the
```

richvdh reviewed (dismissed):
still a few changes needed from the previous review.

babolivier committed:
Incorporate review

babolivier committed:
Update synapse/third_party_rules/access_rules.py

Co-authored-by: Richard van der Hoff <1389908+richvdh@users.noreply.github.com>

richvdh reviewed (approved):
lgtm!
